examples/routing_service/shmem_udp_gateway/README.md

```diff
@@ -0,0 +1,179 @@
+# Shared Memory / UDP Gateway
+
+This example contains the necessary files to run the RTI Routing Service
+example from the "Breaking through Hospital IT Silos: The Top 3 Challenges
+to Overcome" blogpost. Routing Service (RS) will help us interfacing
+between the shared memory domain and the UDP domain. Applications local to a
+host will communicate with each other over SHMEM. RS will be the gateway for
+communication between local and remote applications. For this, RS will create
+a SHMEM DP (no UDP ports) and a UDP DP (with the 3 default UDP ports). If
+multicast is disabled, it will only open 2 UDP ports.
+
+In this example, you will use RTI DDS Ping as a pub/sub example application.
+Along with, Routing Service you will be able to create a gateway between a
+Shared Memory domain and a UDP domain that uses only 3 ports. There is also a
+second configuration option using RTI Cloud Discovery Service to use Unicast.
+
+## Environment variables
+
+For your convenience, there are 2 scripts to set up environment variables:
+
+- Linux: *variables.sh*
+- Windows: *variables.bat*
+
+These are the variables they contain, which you should modify according to your
+system:
+
+-   **NDDSHOME**: path to the installation of RTI Connext Professional.
+-   **NDDS_QOS_PROFILES**: path to the *qos.xml* file containing QoS profiles.
+-   **SHMEM_DOMAIN**: the domain for the applications using SHMEM.
+-   **UDP_DOMAIN**: the domain for the applications using UDP (mainly the
+DomainParticipant of RTI Routing Service).
+-   **APPS**: total number of DomainParticipants on the localhost. By default,
+Connext will try to reach out to the first 5 created applications on SHMEM,
+therefore, we need to increase that number if there are more than 5
+applications.
+-   **CDS_IP_ADDRESS**: the IP address of the host that contains RTI Cloud
+Discovery Service.
+-   **CDS_PORT**: the UDP port that CDS will use.
+
+On Linux, you can use the environment variables by sourcing the file:
+
+```bash
+source variables.sh
+```
+
+On Windows, simply run it:
+
+```bash
+> variables.bat
+```
+
+For convenience, the rest of the README will use the Linux variable sign ($)
+instead of the Windows variable signs (%%).
+
+## Multicast example
+
+You will need 3 terminals to run this example.
+
+1.  On terminal 1, source the variables script and run an RTI DDS Ping
+publisher on SHMEM acting as a local publisher:
+
+    ```bash
+    source variables.sh
```

Comment:
I'd remove the dependency on this file altogether. You could do the following:

```bash
<install_dir>/bin/rtiddsping -pub -domain 0 \
    -qosFile <example_dir>/qos.xml \
    -qosProfile "example_library::smem_profile"

<install_dir>/bin/rtiddsping -sub -domain 1 \
    -qosFile <example_dir>/qos.xml \
    -qosProfile "example_library::multicast"
```
You could simply point to the configuration tweaks in the XML if you want to change APPs and CDS port and IP address.

As for Routing Service, you can provide the value of environment variables via command line argument as follows, which I think it's pretty neat  (this is what we do in our Docker images):

```bash
<install_dir>/bin/rtiroutingservice  \
    -cfgFile <example_dir>/RS_config_multicast.xml \
    -cfgName gateway_SHMEM_and_UDP \
    -DSHMEM_DOMAIN=0 =DUDP_DOMAIN=1
```

Same with Cloud Discovery Service:

````bash
<install_dir>/bin/rticlouddiscoveryservice -cfgFile CDS_config.xml -cfgName cds_all_domains_udpv4 \
    -DSHMEM_DOMAIN=0 =DUDP_DOMAIN=1
````

Reply:
@fgaranda I don't fully agree with your first recommendation. Both ways definitely work, but using environment variables provides flexibility since the customer doesn't need to hardcode magic numbers. At the end of the day, we need to show the customer how to use best practices, and using environment variables is one of them. Specially since every host in the network in this scenario will need different values for those environment variables.

@fgaranda wrt your second recommendation, I agree that using -D is a cool feature, but using "centralized" environment variables in a file gives the user full control. They can use the same method for both their applications (XML QoS) and their Infrastructure Services (Routing Service and CDS). Whereas hardcoding values on XML and using -D for IS are 2 different methods to do the same thing.